### DISCUSSION TITLE
[rfc] layers: custom CSP merger

### DISCUSSION THREAD

**Discussion by dargmuesli** [Baroshem/nuxt-security#198] [Ideas]
Hey, I'm thinking about CSP and Nuxt layers... I suppose that currently in nuxt-security the headers of different layers would be merged using the standard defu strategy (as with all Nuxt config).

I'm thinking about ways to widen the possibilities there. Imagine one layer defining 'none' and the next layer defining a value. That would require 'none' to be removed. Then another layer might come which wishes to override the previously set value back to 'none'. I think that's currently not possible, but would be cool if it was! A custom merger would surely be needed.

But before I hop into custom code I hereby want to make those thoughts transparent as there might be other good ideas about that too! What comes to your mind?

**Comment by dargmuesli** [comment]
I think in the end the most plausible solution is to let each layer define a function which receives the result of all earlier layers as input parameter and freely do any changes to that. I think of something like `nuxt-i18n`'s `i18n.config.ts` (https://github.com/nuxt-modules/i18n/pull/1948).
Putting constraints on that with static object syntax will in any case be limiting.

**Reply by dargmuesli** [reply]
Some links that might be worth a look at implementation:
- https://www.npmjs.com/package/csp-header
- https://www.npmjs.com/package/csp_evaluator
- https://nuxt.com/docs/guide/going-further/layers#multi-layer-support-for-nuxt-modules

**Reply by Baroshem** [reply]
Hey @dargmuesli 

Thank you for this feature request and starting a discussion about it!

Let's wait for some more comments before we start working on it :)

**Reply by Baroshem** [reply]
Hey @dargmuesli 

Would you be interested in developing such functionality? I am right now moving into more active development for the 1.0.0 version of the module and this feature would suit there really well :)

I think the solution created in nuxt-i18n could work well for us as well.

**Reply by dargmuesli** [reply]
I am, but it'll take me a few days to get to this I think :raised_hands:

**Reply by Baroshem** [reply]
No worries! Take your time and let me know if there is anything I can help with :)

**Reply by dargmuesli** [reply]
I now think there is no need for a custom CSP merger because a Nuxt layer can use a(n inline) module to make changes to the `security` configuration like to remove the default `'none'` on `object-src`:

```ts
export default defineNuxtConfig(
  {
    extends: 'some-layer-adding-nuxt-security',
    modules: [
      (_options, nuxt) => {
        const nuxtConfigSecurity = nuxt.options.security
        if (
          typeof nuxtConfigSecurity.headers !== 'boolean' &&
          nuxtConfigSecurity.headers.contentSecurityPolicy &&
          typeof nuxtConfigSecurity.headers.contentSecurityPolicy !==
            'boolean' &&
          typeof nuxtConfigSecurity.headers.contentSecurityPolicy !==
            'string' &&
          nuxtConfigSecurity.headers.contentSecurityPolicy['object-src']
        ) {
          nuxtConfigSecurity.headers.contentSecurityPolicy['object-src'] =
            nuxtConfigSecurity.headers.contentSecurityPolicy[
              'object-src'
            ].filter((x) => x !== "'none'")
        }
        console.log(nuxt.options.security)
      },
    ],
  }
)
```

That's for the v1 rc. A little bit verbose due to the type checks. Let's see how types end up being defined for the v1 release.

**Reply by Baroshem** [reply]
I see. It makes complete sense.

To not let this info disappear, maybe we could add it to the new documenation somewhere? Maybe in the configuration section below the configuration options?

**Reply by dargmuesli** [reply]
Here you go: https://github.com/Baroshem/nuxt-security/pull/226 :rocket:

**Reply by Baroshem** [reply]
My Man! Great work, I will review it in the next few hours :)
